skorch/lr_scheduler.py

```diff
@@ -0,0 +1,132 @@
+        self.policy = policy
+        self._lr_scheduler = None
+        self._init_kwargs = kwargs
+
+    def on_train_begin(self, net, **kwargs):
+        self._lr_scheduler = self._get_scheduler(net, **self._init_kwargs)
+
+    def on_epoch_begin(self, net, **kwargs):
+        epoch = len(net.history)-1
+        if self.policy == 'reduce_plateau':
+            metrics = previous_epoch_train_loss_score(net) if epoch else np.inf
+            self._lr_scheduler.step(metrics, epoch)
+        else:
+            self._lr_scheduler.step(epoch)
+
+    def _get_scheduler(self, net, **kwargs):
+        optimizer = net.optimizer_
+        last_epoch = kwargs.get('last_epoch', -1)
+        if self.policy == 'lambda':
+            lr_lambda = kwargs.get('lambda', lambda x: 1e-1)
+            return LambdaLR(optimizer, lr_lambda, last_epoch)
+
+        if self.policy == 'step':
+            step_size = kwargs.get('step_size', 30)
+            gamma = kwargs.get('gamma', 1e-1)
+            return StepLR(optimizer, step_size, gamma, last_epoch)
+
+        if self.policy == 'multi_step':
+            milestone = kwargs.get('milestone', [30, 90])
+            gamma = kwargs.get('gamma', 1e-1)
+            return MultiStepLR(optimizer, milestone, gamma, last_epoch)
+
+        if self.policy == 'exponential':
+            gamma = kwargs.get('gamma', 1e-1)
+            return ExponentialLR(optimizer, gamma, last_epoch)
+
+        if self.policy == 'reduce_plateau':
+            mode = kwargs.get('mode', 'min')
+            factor = kwargs.get('factor', 1e-1)
+            patience = kwargs.get('patience', 10)
+            verbose = kwargs.get('verbose', False)
+            threshold = kwargs.get('threshold', 1e-4)
+            threshold_mode = kwargs.get('threshold_mode', 'rel')
+            cooldown = kwargs.get('cooldown', 0)
+            min_lr = kwargs.get('min_lr', 0)
+            eps = kwargs.get('eps', 1e-8)
+            return ReduceLROnPlateau(
+                optimizer, mode, factor, patience, verbose, threshold,
+                threshold_mode, cooldown, min_lr, eps
+            )
+
+        if self.policy == 'warm_restart':
+            min_lr = kwargs.get('min_lr', 1e-6)
+            max_lr = kwargs.get('max_lr', 0.05)
+            base_period = kwargs.get('base_period', 10)
+            period_mult = kwargs.get('period_mult', 2)
+            return WarmRestartLR(
+                optimizer, min_lr, max_lr, base_period,
+                period_mult, last_epoch
+            )
+
+class WarmRestartLR(_LRScheduler):
+    """Sets the learning rate of each parameter group according to
+    stochastic gradient descent with warm restarts (SGDR) policy. The
+    policy simulates periodic warm restarts of SGD, where in each restart the
+    learning rate is initialize to some value and is scheduled to decrease,
+    as described in the paper
+    `"Stochastic Gradient Descent with Warm Restarts"
+    <https://arxiv.org/pdf/1608.03983.pdf>`_
+    Args:
```

Comment:
Can you format this in the numpy doc style so that it gets rendered correctly in the online documentation? (See [section 4](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections)).

Reply:
Sure.